### PR TITLE
Catch up with missing preimages regardless of nomination status

### PR DESCRIPTION
**Pull request by Geod24**
Previously, we were only catching up with pre-images we've seen as missing
through side effects of the `isInvalidPreimageRootReason`.
This `private` member of the `Ledger` was then cleared in the `NetworkManager`.

Instead of relying on this tight coupling and hard to follow logic,
and as we want to try to catch-up with all missing pre-images anyways,
we can just add a task in `Validator` which behaves similarly to the block catchup.

### CODE REVIEW

File: source/agora/node/Validator.d [hunk 1]
```diff
@@ -248,6 +251,76 @@ public class Validator : FullNode, API
...
+
+        auto missing = validators.filter!(en => en.preimage.height < next_height).array();
+        if (!missing.length)
+            return;
+
+        log.warn("Currently missing pre-images for next height ({}): {}",
+                 next_height, missing);
+
+        auto query = this.network.peers[]
+            .map!(peer => peer.client.getPreimages(next_height, next_height + 100));
```
Comment (hewison-chris) [inline source/agora/node/Validator.d]: Why limit to pre-images 100 higher ? Any future pre-image will do.

File: source/agora/test/PreimageSharing.d [hunk 2]
```diff
@@ -20,18 +20,18 @@ import agora.consensus.data.PreImageInfo;
...
-    /// GET: /preimages_for_enroll_keys
-    public override PreImageInfo[] getPreimagesForEnrollKeys (Set!Hash enroll_keys = Set!Hash.init) @safe nothrow
-    {
-        return TestValidatorNode.getPreimagesForEnrollKeys(enroll_keys);
-    }
+    ///
+    public override void onPreImageRevealTimer () @safe {}
+
+    /// To be extra sure, we also disable receiving a pre-image
+    /// so that node node may gossip them
```
Comment (hewison-chris) [inline source/agora/test/PreimageSharing.d]:
```suggestion
    /// so that the node may gossip them
```

**Review by hewison-chris — APPROVED** [review]
LGTM

File: source/agora/node/Validator.d (hunk 1 above)
Comment (Geod24) [reply source/agora/node/Validator.d]: There's no way to express this in the API. Perhaps I should just call it with `Height.max`

Comment (Geod24) [reply source/agora/node/Validator.d]: Thanks, new version looks much better.

**Review by linked0 — APPROVED** [review]
Nice!

**Comment by Geod24** [comment]
> Thanks, new version looks much better.

This should have been preceded by a `git push`. Done  now.

**Review by hewison-chris — APPROVED** [review]
LGTM

**Review by kchulj — APPROVED** [review]
